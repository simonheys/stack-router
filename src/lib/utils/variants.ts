export type StackPresentation = "default" | "modal" | "modal-sheet"

export const variants = {
  enter: {
    filter: "brightness(1)",
    translateX: "100%",
    translateY: 0,
    boxShadow: "0 0 0 0 rgba(0, 0, 0, 0)",
    transformOrigin: "50% 0",
  },
  active: {
    filter: "brightness(1)",
    translateX: 0,
    translateY: 0,
    marginTop: 0,
    scale: 1,
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0,
    boxShadow: "0 0 20px 5px rgba(0, 0, 0, 0.3)",
    transformOrigin: "50% 0",
  },
  activeModalSheet: {
    filter: "brightness(1)",
    translateX: 0,
    translateY: 0,
    marginTop: "30px",
    scale: 1,
    borderTopLeftRadius: "4px",
    borderTopRightRadius: "4px",
    boxShadow: "0 0 20px 5px rgba(0, 0, 0, 0.3)",
    transformOrigin: "50% 0",
  },
  exit: {
    filter: "brightness(0.3)",
    translateX: "-30%",
    translateY: 0,
    transformOrigin: "50% 0",
  },
  stacked: {
    filter: "brightness(0.5)",
    translateX: 0,
    translateY: 0,
    marginTop: "10px",
    scale: 0.9,
    borderTopLeftRadius: "4px",
    borderTopRightRadius: "4px",
    transformOrigin: "50% 0",
  },
  exitStacked: {
    filter: "brightness(0)",
    translateX: 0,
    translateY: 0,
    scale: 0.8,
    borderTopLeftRadius: "4px",
    borderTopRightRadius: "4px",
    transformOrigin: "50% 0",
  },
  exitModal: {
    filter: "brightness(1)",
    translateX: 0,
    translateY: "100%",
    boxShadow: "0 0 0 0 rgba(0, 0, 0, 0)",
    borderTopLeftRadius: "4px",
    borderTopRightRadius: "4px",
    transformOrigin: "50% 0",
  },
}

type VariantKey = keyof typeof variants

export interface IGetVariants {
  isEntering: boolean
  isExiting: boolean
  isPush: boolean
  isPop: boolean
  presentationFrom: StackPresentation | undefined
  presentationTo: StackPresentation
}

export const getPresentationVariants = ({
  isEntering,
  isExiting,
  isPush,
  isPop,
  presentationFrom,
  presentationTo,
}: IGetVariants): [VariantKey, VariantKey, VariantKey] => {
  let enter: VariantKey = "enter"
  let active: VariantKey = "active"
  let exit: VariantKey = "exit"

  if (!isEntering && !isExiting) {
    enter = "stacked"
    exit = "stacked"
    return [enter, active, exit]
  }

  if (isEntering) {
    if (isPush) {
      if (presentationTo === "modal") {
        enter = "exitModal"
      }
      if (presentationTo === "modal-sheet") {
        enter = "exitModal"
        active = "activeModalSheet"
      }
      if (
        (presentationFrom === "modal" || presentationFrom === "modal-sheet") &&
        presentationTo === "default"
      ) {
        enter = "enter"
      }
    } else if (isPop) {
      enter = "exit"
      if (presentationTo === "modal-sheet") {
        active = "activeModalSheet"
      }
      if (presentationFrom === "modal" || presentationFrom === "modal-sheet") {
        enter = "stacked"
      }
      if (
        presentationFrom === "modal-sheet" &&
        presentationTo === "modal-sheet"
      ) {
        exit = "exitModal"
      }
    }
  } else if (isExiting) {
    if (isPush) {
      if (presentationTo === "modal" || presentationTo === "modal-sheet") {
        exit = "stacked"
      } else if (
        (presentationFrom === "modal" || presentationFrom === "modal-sheet") &&
        presentationTo === "default"
      ) {
        exit = "exit"
      }
    }
    if (isPop) {
      exit = "enter"
      if (
        presentationFrom === "default" &&
        (presentationTo === "modal" || presentationTo === "modal-sheet")
      ) {
        exit = "enter"
      } else if (
        (presentationFrom === "modal" || presentationFrom === "modal-sheet") &&
        presentationTo === "default"
      ) {
        exit = "exitModal"
      } else if (
        presentationFrom === "modal-sheet" &&
        presentationTo === "modal-sheet"
      ) {
        exit = "exitModal"
      }
    }
  }

  return [enter, active, exit]
}
