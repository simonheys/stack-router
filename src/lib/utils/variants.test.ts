import { StackPresentation, getPresentationVariants } from "./variants"

export const dumpVariants = () => {
  const flags = [true, false]
  const presentations: StackPresentation[] = ["default", "modal", "modal-sheet"]
  let result = ""
  for (const presentationFrom of presentations) {
    for (const presentationTo of presentations) {
      for (const isEntering of flags) {
        const isExiting = !isEntering
        for (const isPush of flags) {
          const isPop = !isPush
          const [enter, active, exit] = getPresentationVariants({
            isEntering,
            isExiting,
            isPush,
            isPop,
            presentationFrom,
            presentationTo,
          })
          result += `${isEntering ? "entering" : "exiting"} ${
            isPush ? "push" : "pop"
          } ${presentationFrom} -> ${presentationTo} [${enter}, ${active}, ${exit}]\r`
        }
      }
    }
  }
  return result
}

describe("variants", () => {
  describe("getPresentationVariants", () => {
    test("getPresentationVariants", () => {
      expect(dumpVariants()).toMatchSnapshot()
    })
  })
})
