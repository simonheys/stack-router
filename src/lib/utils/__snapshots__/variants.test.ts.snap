// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variants getPresentationVariants getPresentationVariants 1`] = `
"entering push default -> default [enter, active, exit]
entering pop default -> default [exit, active, exit]
exiting push default -> default [enter, active, exit]
exiting pop default -> default [enter, active, enter]
entering push default -> modal [exitModal, active, exit]
entering pop default -> modal [exit, active, exit]
exiting push default -> modal [enter, active, stacked]
exiting pop default -> modal [enter, active, enter]
entering push default -> modal-sheet [exitModal, activeModalSheet, exit]
entering pop default -> modal-sheet [exit, activeModalSheet, exit]
exiting push default -> modal-sheet [enter, active, stacked]
exiting pop default -> modal-sheet [enter, active, enter]
entering push modal -> default [enter, active, exit]
entering pop modal -> default [stacked, active, exit]
exiting push modal -> default [enter, active, exit]
exiting pop modal -> default [enter, active, exitModal]
entering push modal -> modal [exitModal, active, exit]
entering pop modal -> modal [stacked, active, exit]
exiting push modal -> modal [enter, active, stacked]
exiting pop modal -> modal [enter, active, enter]
entering push modal -> modal-sheet [exitModal, activeModalSheet, exit]
entering pop modal -> modal-sheet [stacked, activeModalSheet, exit]
exiting push modal -> modal-sheet [enter, active, stacked]
exiting pop modal -> modal-sheet [enter, active, enter]
entering push modal-sheet -> default [enter, active, exit]
entering pop modal-sheet -> default [stacked, active, exit]
exiting push modal-sheet -> default [enter, active, exit]
exiting pop modal-sheet -> default [enter, active, exitModal]
entering push modal-sheet -> modal [exitModal, active, exit]
entering pop modal-sheet -> modal [stacked, active, exit]
exiting push modal-sheet -> modal [enter, active, stacked]
exiting pop modal-sheet -> modal [enter, active, enter]
entering push modal-sheet -> modal-sheet [exitModal, activeModalSheet, exit]
entering pop modal-sheet -> modal-sheet [stacked, activeModalSheet, exitModal]
exiting push modal-sheet -> modal-sheet [enter, active, stacked]
exiting pop modal-sheet -> modal-sheet [enter, active, exitModal]
"
`;
